pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,137 @@
+import torch
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.indep_messenger import CondIndepStackFrame
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
```

Comment:
@fritzo - This is ready for another round of review. 😄The important change is confined to this class. In HMC we are calling the constructor once to populate the children nodes and compute which dims should be evaluated/enumerated out. When we call `log_prob` subsequently, we reuse this computed metadata to aggregate the log prob terms from the leaf nodes upwards.